Source/SwiftLintFramework/Rules/Lint/PrivateSwiftUIStatePropertyRule.swift

```diff
@@ -0,0 +1,99 @@
+import SwiftSyntax
+
+/// Rule to require that any state properties in SwiftUI be declared as private.
+/// State properties should only be accessible from inside the View's body, or from methods called by it
+struct PrivateSwiftUIStatePropertyRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule {
+    var configuration = SeverityConfiguration(.warning)
+
+    static let description = RuleDescription(
+        identifier: "private_swiftui_state",
+        name: "Private SwiftUI @State properties",
+        description: "SwiftUI state properties should be private",
+        kind: .lint,
+        nonTriggeringExamples: [
+            Example(
+                """
+                struct ContentView: View {
+                    @State private var isPlaying: Bool = false
+                }
+                """
+            ),
+            Example(
+                """
+                struct ContentView: View {
+                    @State fileprivate var isPlaying: Bool = false
+                }
+                """
+            ),
+            Example(
+                """
+                struct ContentView: View {
+                    var isPlaying = false
+                }
+                """
+            ),
+            Example(
+                """
+                struct ContentView: View {
+                    @StateObject var foo = Foo()
+                }
+                """
+            )
+        ],
+        triggeringExamples: [
+            Example(
+                """
+                struct ContentView: View {
+                    @State var isPlaying: Bool = false
+                }
+                """
+            )
+        ])
+
+    init() {}
+
+    func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
+
+private extension PrivateSwiftUIStatePropertyRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ node: MemberDeclListItemSyntax) {
+            guard
+                let decl = node.decl.as(VariableDeclSyntax.self),
+                    decl.attributes?.hasStateAttribute == true,
+                    decl.modifiers?.isPrivateOrFilePrivate != true
+            else {
+                return
+            }
+
+            violations.append(decl.letOrVarKeyword.positionAfterSkippingLeadingTrivia)
+        }
+    }
+}
+
+private extension AttributeListSyntax {
+    var hasStateAttribute: Bool {
+        contains { attr in
+            guard let stateAttr = attr.as(CustomAttributeSyntax.self),
+                  let identifier = stateAttr.attributeName.as(SimpleTypeIdentifierSyntax.self) else {
+                return false
+            }
+
+            return identifier.name.text == "State"
+        }
+    }
+}
+
+private extension ModifierListSyntax {
```

Comment:
Please check if some of these extensions are already defined in `SwiftSyntax+SwiftLint.swift`. If so, please use those. If not, think about integrating yours with the existing ones.